[WIP] Add support for viewing a message in the web browser

preetmishra opened the pull request:
Thanks to @punchagan for the initial work in #397. :+1: 

Unlike the original PR, which narrowed every message to All Messages in the web app, this opens a message in `stream/topic/near` if it belongs to a stream or `pm-with/near` narrow if it belongs to a PM/huddle (see [discussion](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/View.20message.20in.20browser/near/908851)).

The original commit has been split to facilitate reviews. Moreover, the test has been improved based on @neiljp's feedback on the original PR.

The logic for generating near message URLs has been borrowed from Zulip's [`zerver/lib/url_encoding.py`](https://github.com/zulip/zulip/blob/master/zerver/lib/url_encoding.py).

Partially fixes #452.

neiljp commented:
@preetmishra We discussed this on czo, but let me know if you need more specific feedback.

preetmishra commented:
@neiljp I have updated the PR based on your suggestions on CZO. :+1: 

I have split the commit from the original PR into three, without any alternations, and added my changes into separate commits.

I will proceed to work on other enhancements that we discussed once the commit structure looks good to you.

preetmishra commented:
Updated to resolve conflicts.

preetmishra commented:
Updated to resolve conflicts.

preetmishra commented:
Updated to resolve conflicts.

neiljp commented on zulipterminal/helper.py:
```diff
@@ -666,3 +667,63 @@ def suppress_output() -> Iterator[None]:
         os.dup2(err, 2)
+
+
+def hash_util_encode(string: str) -> str:
```
Let's use a different file for these, since we could combine these with the URL deconstruction too and/or put them into the API repo like that other method in future?

neiljp commented on zulipterminal/core.py:
```diff
@@ -298,15 +300,16 @@ def show_all_mentions(self, button: Any) -> None:
-    def view_in_browser(self, message_id: int) -> None:
-        url = '{}#narrow/near/{}'.format(self.model.server_url, message_id)
+    def view_in_browser(self, message: Message) -> None:
+        # Truncate extra '/' from the server url.
+        self.url = near_message_url(self.model.server_url[:-1], message)
```
It's good to use the pre-existing code here, though using `urljoin` in the new code would mean we don't need to worry about this. I'm guessing the server code avoids this since it knows the url exactly and relies on tests, which we should too really :)

neiljp commented on zulipterminal/core.py:
```diff
@@ -298,15 +300,16 @@ def show_all_mentions(self, button: Any) -> None:
         with suppress_output():
             # Suppress anything on stdout or stderr when opening the browser
-            webbrowser.open(url)
+            webbrowser.open(self.url)
```
Why do we need the `self`?

neiljp commented on zulipterminal/core.py:
```diff
@@ -298,15 +300,16 @@ def show_all_mentions(self, button: Any) -> None:
 
-    def view_in_browser(self, message_id: int) -> None:
-        url = '{}#narrow/near/{}'.format(self.model.server_url, message_id)
+    def view_in_browser(self, message: Message) -> None:
```
This change breaks the other use of this method - but no tests break?

zee-bit commented:
@preetmishra We recently merged #854, which centralizes the logic of generating `near_message_url()` to a new module `server_url.py`(along with the tests in a corresponding `test_*` file). This should simplify this PR to a great extent. :)

(Let me know if you are busy and want me to take this further)

preetmishra commented:
@zee-bit, thanks for the update! I will update the PR soon.

preetmishra commented:
@zee-bit Hey! I am a bit busy with work. However, I rebased my local branch onto master. Feel free to take it further. Best of luck!

neiljp commented:
@preetmishra As you know this was replaced by the now-merged #991 so I'm closing this :tada:
